src/fsharp/TypeChecker.fs

```diff
@@ -6109,18 +6110,19 @@ and TcRecordConstruction cenv overallTy env tpenv optOrigExpr objTy fldsList m =
               (fname,fieldExpr),tpenv)
               
     // Add rebindings for unbound field when an "old value" is available 
-    let oldFldsList = 
+    // Effect order: mutable fields may get modified by other bindings... 
+    let oldFldsList, wrap = 
         match optOrigExpr with
-        | None -> []
-        | Some (_,_,oldve') -> 
-               // When we have an "old" value, append bindings for the unbound fields. 
-               // Effect order - mutable fields may get modified by other bindings... 
-               let fieldNameUnbound nom = List.forall (fun (name,_) -> name <> nom) fldsList
-               fspecs 
-               |> List.choose (fun rfld -> 
+        | None -> [], id
+        | Some (_,_,oldve) -> 
+            let wrap,oldveaddr = mkExprAddrOfExpr cenv.g tycon.IsStructOrEnumTycon false NeverMutates oldve None m
+            let fieldNameUnbound nom = List.forall (fun (name,_) -> name <> nom) fldsList
+            let flds = 
+                fspecs |> List.choose (fun rfld -> 
                     if fieldNameUnbound rfld.Name && not rfld.IsZeroInit 
-                    then Some(rfld.Name, mkRecdFieldGet cenv.g (oldve',tcref.MakeNestedRecdFieldRef rfld,tinst,m))
+                    then Some(rfld.Name, mkRecdFieldGetViaExprAddr (oldveaddr,tcref.MakeNestedRecdFieldRef rfld,tinst,m))
                     else None)
```

Comment:
@TIHan This is problably the most important fix.  Previously ``{ r with x = 1; y = 2 }`` was becoming ``{ x=1; y=2; z= (COPY-OF-R).z; w = (COPY-OF-R).w }``  That is each ``old`` field induced a copy of ``r``.  

Now, we take the address of ``r`` once (making a silent copy in order to take the address if needed).  Then the record expression becomes ``{ x=1; y=2; z= (ADDRESS-OF-R)->z; w = (ADDRESS-OF-R)->w }``

But more testing should be added around this, partly because the difference might be seen in quotations.  Also a codegen EmittedIL test should probably be added for this case